### PR TITLE
LibJS: General parser fixes and partial unicode support

### PR DESCRIPTION
**Pull request by davidot**
Fixes a bunch of test262-parser tests and has no regressions in test262.

The parser is quite messy (even more now) but I plan to try to kind of fix that in a future PR since bindings and variable scoping checks are getting very complicated and duplicated all over the place.

### CODE REVIEW

File: Userland/Libraries/LibJS/Parser.cpp [hunk 1]
```diff
@@ -2585,15 +2678,48 @@ bool Parser::match_statement() const
...
+
+    auto& mutable_me = const_cast<Parser&>(*this);
+
+    mutable_me.save_state();
+
+    ScopeGuard state_rollback = [&] {
+        mutable_me.load_state();
+    };
+
+    mutable_me.consume(TokenType::Let);
```
Comment (davidot) [inline Userland/Libraries/LibJS/Parser.cpp:2704]: Using `const_cast` is not great but we don't have peek so this was the best way I could think of.
One alternative is creating a full new lexer starting at the let.

**Comment by davidot** [comment]
The CI failure is a timeout

File: Userland/Libraries/LibJS/Lexer.cpp [hunk 2]
```diff
@@ -186,6 +189,26 @@ void Lexer::consume()
...
+        }
+
+        VERIFY(char_size > 1);
+        --char_size;
+
+        m_position += char_size;
+        if (did_reach_eof())
+            return;
+
+        m_line_column++;
```
Comment (davidot) [inline Userland/Libraries/LibJS/Lexer.cpp:211]: Not great to have this logic here but there was no spot I could find which just had this.

File: Userland/Libraries/LibJS/Lexer.cpp [hunk 3]
```diff
@@ -310,21 +333,55 @@ bool Lexer::is_line_terminator() const
...
+    return (m_current_char & 128) != 0;
+}
+
+u32 Lexer::current_codepoint() const
+{
+    static constexpr const u32 REPLACEMENT_CHARACTER = 0xFFFD;
+    if (m_position == 0)
+        return REPLACEMENT_CHARACTER;
+    Utf8View utf_8_view {m_source.substring_view(m_position - 1)};
+    return *utf_8_view.begin();
```
Comment (davidot) [inline Userland/Libraries/LibJS/Lexer.cpp:354]: Using `Utf8View` here since it can decode but I don't know if this is the best way to do this

File: Userland/Libraries/LibJS/Lexer.cpp [hunk 4]
```diff
@@ -9,8 +9,14 @@
 #include <AK/CharacterTypes.h>
 #include <AK/Debug.h>
 #include <AK/HashMap.h>
+#include <AK/Utf8View.h>
+#include <LibUnicode/CharacterTypes.h>
 #include <stdio.h>
 
+#ifdef ENABLE_UNICODE_DATA
+#   include <LibUnicode/UnicodeData.h>
+#endif
```
Comment (davidot) [inline Userland/Libraries/LibJS/Lexer.cpp:18]: Copied this pattern from some other place. Is this how we are supposed to guard any unicode data usage? (cc @trflynn89 perhaps)

Comment (trflynn89) [reply Userland/Libraries/LibJS/Lexer.cpp:18]: Oh, hmm, so I was trying to keep the `#ifdef` mess completely hidden inside LibUnicode's .cpp files, but I see that doesn't quite work when callers want to refer to the generated enums by their values directly...

I'll think about how to make this nicer for user cases like this. In the meantime, to avoid the `#ifdef` spaghetti, you could do something like:
```c++
bool Lexer::is_whitespace() const
{
    ...
    
    static auto space_separator = Unicode::general_category_from_string("Space_Separator"sv);
    if (space_separator.has_value())
        return Unicode::code_point_has_general_category(codepoint, *space_separator);
    return false;
}
```

Comment (davidot) [reply Userland/Libraries/LibJS/Lexer.cpp:18]: Yeah seems hard to avoid if you want to use the enum directly.
I'll change it to that pattern for now then. Thanks!

Comment (trflynn89) [reply Userland/Libraries/LibJS/Lexer.cpp:18]: Oh, semi unrelated, but IIRC at some point someone went through the code base and standardized "code point" being two words. So `codepoint` var names should be `code_point` to match that :)

Comment (davidot) [reply Userland/Libraries/LibJS/Lexer.cpp:18]: Hmm seems some of the categories are not available for both `ID_Start` and `ID_Continue`. Am I doing something wrong?

Comment (IdanHo) [reply Userland/Libraries/LibJS/Lexer.cpp:18]: I had the same direct access issue when i was briefly trying to update libgfx's unicode bidi implementation to use LibUnicode, so +1 for that

Comment (trflynn89) [reply Userland/Libraries/LibJS/Lexer.cpp:18]: `ID_Start` and `ID_Continue` are properties, so use `Unicode::property_from_string` for those (if you aren't already)

Comment (davidot) [reply Userland/Libraries/LibJS/Lexer.cpp:18]: > Oh, semi unrelated, but IIRC at some point someone went through the code base and standardized "code point" being two words. So `codepoint` var names should be `code_point` to match that :)

Even my commit was inconsistent so was already fixing it. Thanks for the good eye

File: Userland/Libraries/LibJS/Lexer.cpp [hunk 5]
```diff
@@ -310,21 +336,69 @@ bool Lexer::is_line_terminator() const
...
 {
-    return is_ascii_alpha(m_current_char) || m_current_char == '_' || m_current_char == '$';
+    if (!is_unicode_character())
+        return is_ascii_alpha(m_current_char) || m_current_char == '_' || m_current_char == '$';
+    auto code_point = current_code_point();
+
+    static auto id_start_category = Unicode::property_from_string("ID_Start"sv);
+    if (id_start_category.has_value())
+        return Unicode::code_point_has_property(code_point, *id_start_category);
+    dbgln("no category!");
```
Comment (trflynn89) [inline Userland/Libraries/LibJS/Lexer.cpp:385]: Should probably get rid of this dbgln? and the one in `is_identifier_middle`

File: Userland/Libraries/LibJS/Parser.cpp [hunk 6]
```diff
@@ -400,6 +400,11 @@ NonnullRefPtr<Statement> Parser::parse_statement(AllowLabelledFunction allow_lab
     case TokenType::Semicolon:
         consume();
         return create_ast_node<EmptyStatement>({ m_state.current_token.filename(), rule_start.position(), position() });
+    case TokenType::Slash:
+    case TokenType::SlashEquals:
+        // Lexer was wrong this is a regexexpr
```
Comment (linusg) [inline Userland/Libraries/LibJS/Parser.cpp:405]: Please write comments more carefully, as whole sentences including punctuation. See https://github.com/SerenityOS/serenity/blob/master/Documentation/CodingStyle.md#comments.

Also, `regexexpr` is weird - the `ex` in `regex` already stands for expression...

File: Userland/Libraries/LibJS/Parser.cpp [hunk 7]
```diff
@@ -906,7 +911,6 @@ NonnullRefPtr<RegExpLiteral> Parser::parse_regexp_literal()
     auto pattern = consume().value();
     // Remove leading and trailing slash.
     pattern = pattern.substring_view(1, pattern.length() - 2);
-
```
Comment (linusg) [inline Userland/Libraries/LibJS/Parser.cpp:865]: Please keep unrelated whitespace changes to a minimum, this seems unnecessary

File: Userland/Libraries/LibJS/Parser.cpp [hunk 8]
```diff
@@ -2092,9 +2106,19 @@ NonnullRefPtr<VariableDeclaration> Parser::parse_variable_declaration(bool for_l
...
                     if (name == "let"sv)
                         syntax_error("Lexical binding may not be called 'let'");
+
+                    // FIXME: Again we do not check everything here see for example the parameter check above.
+                    //        However a more sustainable solution should be used since these checks are now spread over multiple sites.
+
+                    for (auto& declaration : declarations) {
+                        declaration.target().visit([&](NonnullRefPtr<Identifier> const& identifier) {
+                                                      if (identifier->string() == name)
+                                                          syntax_error(String::formatted("Identifier '{}' has already been declared", name)); },
```
Comment (linusg) [inline Userland/Libraries/LibJS/Parser.cpp:2168]: What's up with the massive indentation here?!

File: Userland/Libraries/LibJS/Parser.cpp (hunk 1 above)
Comment (linusg) [reply Userland/Libraries/LibJS/Parser.cpp:2776]: just don't mark the method `const` if it isn't?

File: Userland/Libraries/LibJS/Parser.h [hunk 9]
```diff
@@ -160,6 +160,8 @@ class Parser {
     bool match_statement() const;
     bool match_export_or_import() const;
     bool match_declaration() const;
+    bool try_match_let_declaration() const;
+
```
Comment (linusg) [inline Userland/Libraries/LibJS/Parser.h:164]: Why are these now separated?

File: Userland/Libraries/LibJS/Lexer.cpp [hunk 10]
```diff
@@ -163,7 +165,6 @@ void Lexer::consume()
                 type = "LINE SEPARATOR";
             else
                 type = "PARAGRAPH SEPARATOR";
-            dbgln("Found a line terminator: {}", type);
```
Comment (linusg) [inline Userland/Libraries/LibJS/Lexer.cpp:166]: The whole point of this block is to print something, why are you removing this?

**Review by linusg — CHANGES_REQUESTED** [review]
`LibJS: Treat yield as an identifier in more non-generator context` is missing an s at the end, and a period in the body.

File: Userland/Libraries/LibJS/Lexer.cpp (hunk 10 above)
Comment (davidot) [reply Userland/Libraries/LibJS/Lexer.cpp:166]: Whoops thought that was my debugging code

**Comment by davidot** [comment]
Think I got everything there. I made the function non const but that propagated to a number of other functions but I guess they are all private any called via non-const so it does not matter much.

File: Userland/Libraries/LibJS/Parser.cpp [hunk 11]
```diff
@@ -1974,7 +2064,17 @@ RefPtr<BindingPattern> Parser::parse_binding_pattern()
     auto pattern = adopt_ref(*new BindingPattern);
     pattern->entries = move(entries);
     pattern->kind = kind;
-    pattern->for_each_bound_name([this](auto& name) { check_identifier_name_for_assignment_validity(name); });
+
+    Vector<StringView> bound_names;
+    pattern->for_each_bound_name([&](auto& name) {
+        if (strict_checks) {
+            if (bound_names.contains_slow(name)) {
+                syntax_error("Duplicate parameter names in bindings");
```
Comment (davidot) [inline Userland/Libraries/LibJS/Parser.cpp:2072]: Style mistake

**Review by linusg — APPROVED** [review]
⭐
